api/src/main/java/com/google/sps/ServiceCreation.java

```diff
@@ -79,4 +113,92 @@ public static Gmail createService() throws IOException, GeneralSecurityException
 
     return service;
   }
+
+  // Custom datastore implementation with FireStore
+  private static class FireDataStore<V extends Serializable> extends AbstractDataStore<V> {
+
+    FireDataStore(DataStoreFactory dataStoreFactory1, String id1) {
+      super(dataStoreFactory1, id1);
+    }
+
+    @Override
+    public DataStore<V> set(String key, V value) throws IOException {
+      final String encoded = Base64.getEncoder().encodeToString(SerializationUtils.serialize(value));
+      Map<String, Object> data = new HashMap<>();
+      data.put(key, encoded);
+      docRef.set(data);
+
+      return this;
+    }
+
+    @Override
+    public V get(String key) throws IOException {
+      try {
+        DocumentSnapshot document = docRef.get().get();
+        if (!document.exists()) return null;
+        final String encoded = document.getString(key);
+        if (encoded == null) {
+          return null;
+        }
+        return SerializationUtils.deserialize(Base64.getDecoder().decode(encoded));
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
```

Comment:
We can wrap the exception with IOException. https://www.baeldung.com/java-wrapping-vs-rethrowing-exceptions#throw-different-exception, see the console log exception there, it actually makes much sense.

@tomatocat